File: PE-common/jvm/src/main/java/nl/adaptivity/messaging/SendableSoapSource.java
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.messaging;

import nl.adaptivity.io.Writable;
import nl.adaptivity.util.activation.Sources;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3.soapEnvelope.Envelope;

import javax.activation.DataSource;
import javax.xml.transform.Source;
import javax.xml.transform.TransformerException;

import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;


public class SendableSoapSource implements ISendableMessage, Writable {

  private final EndpointDescriptor mDestination;

  private final Source mMessage;

  private final Map<String, DataSource> mAttachments;

  public SendableSoapSource(final EndpointDescriptor destination, final Source message) {
    this(destination, message, Collections.<String,DataSource>emptyMap());
  }

  public SendableSoapSource(final EndpointDescriptor destination, final Source message, final Map<String, DataSource> attachments) {
    mDestination = destination;
    mMessage = message;
    mAttachments = attachments;
  }

  @Override
  public EndpointDescriptor getDestination() {
    return mDestination;
  }

  @Nullable
  @Override
  public String getMethod() {
    return null;
  }

  @NotNull
  @Override
  public Collection<? extends IHeader> getHeaders() {
    return Collections.emptyList();
  }

  @NotNull
  @Override
  public Writable getBodySource() {
    return mMessage==null ? null : this;
  }

  public Reader getBodyReader() {
    return Sources.toReader(mMessage);
  }

  @Override
  public String getContentType() {
    return Envelope.MIMETYPE;
  }

  @Override
  public void writeTo(final Writer destination) throws IOException {
    try {
      Sources.writeToWriter(mMessage, destination);
    } catch (TransformerException e) {
      throw new IOException(e);
    }
  }

  @Override
  public Map<String, DataSource> getAttachments() {
    return mAttachments;
  }

}

File: PE-common/src/jvmMain/kotlin/nl/adaptivity/messaging/SendableSoapSource.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.messaging

import nl.adaptivity.io.Writable
import nl.adaptivity.util.activation.toReader
import nl.adaptivity.util.activation.writeToWriter
import org.w3.soapEnvelope.Envelope
import java.io.IOException
import java.io.Reader
import java.io.Writer
import javax.activation.DataSource
import javax.xml.transform.Source


class SendableSoapSource @JvmOverloads constructor(
    override val destination: EndpointDescriptor,
    private val message: Source,
    override val attachments: Map<String, DataSource> = emptyMap()
                                                  ) : ISendableMessage, Writable {

    override val method: String?
        get() = null

    override val headers: Collection<ISendableMessage.IHeader>
        get() = emptyList()

    override val bodySource: Writable
        get() = this

    override val bodyReader: Reader
        get() = message.toReader()

    override val contentType: String
        get() = Envelope.MIMETYPE

    @Throws(IOException::class)
    override fun writeTo(destination: Writer) {
        message.writeToWriter(destination)
    }

}
